Full name: Nguyễn Quốc Huy

Student ID: 20127188

# HW2: Parallel Execution in CUDA

**To compile your file, you can use this command:** \
`nvcc tên-file.cu -o tên-file-chạy` \
***You can use Vietnamese to anwser the questions***

## Question 1A

In [14]:
!nvcc HW2_P1.cu -o HW2_P1

In [15]:
!./HW2_P1

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835398144 bytes
****************************

Input size: 16777217
Host time: 52.584415 ms

Kernel 1
Grid size: 8193, block size: 1024
Kernel time = 2.603776 ms
CORRECT :)

Kernel 2
Grid size: 8193, block size: 1024
Kernel time = 2.684768 ms
CORRECT :)

Kernel 3
Grid size: 8193, block size: 1024
Kernel time = 1.280896 ms
CORRECT :)


We can rely on the results to see the calculation time of Kernel 3 > Kernel 1 > Kernel 2. This can be understood simply because Kernel 3 uses blockdim/stride and the blocks are consecutive so it will be easy to access. From there, optimize time. Kernel 1 is faster than Kernel 2 because although Kernel 2 is used to reduce divergence, it does more calculations and takes more time.

In [16]:
!nvprof ./HW2_P1

==24378== NVPROF is profiling process 24378, command: ./HW2_P1
**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835398144 bytes
****************************

Input size: 16777217
Host time: 43.348160 ms

Kernel 1
Grid size: 8193, block size: 1024
Kernel time = 2.648704 ms
CORRECT :)

Kernel 2
Grid size: 8193, block size: 1024
Kernel time = 2.730208 ms
CORRECT :)

Kernel 3
Grid size: 8193, block size: 1024
Kernel time = 1.327456 ms
CORRECT :)
==24378== Profiling application: ./HW2_P1
==24378== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   86.64%  42.828ms         3  14.276ms  14.070ms  14.410ms  [CUDA memcpy HtoD]
                    5.45%  2.6940ms         1  2.6940ms  2.6940ms  2.6940ms  reduceBlksKernel2(int*, int*, int)
                    5.27%  2.6055ms         1  2.6055ms  2.6055ms  2.6055ms  reduceBlksKernel1(int*, 

**Memory Transfers Dominant:**

The majority of GPU time is spent on memory transfers from the host to the device.
Reducing the size of data transferred or optimizing memory transfer patterns may help improve overall performance.

**Kernel Execution Times:**

The kernel execution times for Kernels 1, 2, and 3 are relatively small compared to memory transfers.
The efficiency of Kernel 3 is evident as it has the lowest execution time among the three kernels.

## Question 1B


In [10]:
!./HW2_P1 1024

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835398144 bytes
****************************

Input size: 16777217
Host time: 45.377537 ms

Kernel 1
Grid size: 8193, block size: 1024
Kernel time = 2.591584 ms
CORRECT :)

Kernel 2
Grid size: 8193, block size: 1024
Kernel time = 2.676768 ms
CORRECT :)

Kernel 3
Grid size: 8193, block size: 1024
Kernel time = 1.275616 ms
CORRECT :)


In [11]:
!./HW2_P1 512

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835398144 bytes
****************************

Input size: 16777217
Host time: 43.587967 ms

Kernel 1
Grid size: 16385, block size: 512
Kernel time = 1.882080 ms
CORRECT :)

Kernel 2
Grid size: 16385, block size: 512
Kernel time = 1.977216 ms
CORRECT :)

Kernel 3
Grid size: 16385, block size: 512
Kernel time = 1.007616 ms
CORRECT :)


In [12]:
!./HW2_P1 256

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835398144 bytes
****************************

Input size: 16777217
Host time: 44.224545 ms

Kernel 1
Grid size: 32769, block size: 256
Kernel time = 1.533952 ms
CORRECT :)

Kernel 2
Grid size: 32769, block size: 256
Kernel time = 1.585152 ms
CORRECT :)

Kernel 3
Grid size: 32769, block size: 256
Kernel time = 0.845760 ms
CORRECT :)


In [13]:
!./HW2_P1 128

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835398144 bytes
****************************

Input size: 16777217
Host time: 49.709057 ms

Kernel 1
Grid size: 65537, block size: 128
Kernel time = 1.346752 ms
CORRECT :)

Kernel 2
Grid size: 65537, block size: 128
Kernel time = 1.373056 ms
CORRECT :)

Kernel 3
Grid size: 65537, block size: 128
Kernel time = 0.746912 ms
CORRECT :)


Block size | Grid size | Num blocks / SM | Occupancy (%) | Kernel time (ms)
--- | --- | --- | --- | ---
1024 | 8193 | 1024/1024 = 1 |  | 2.591584
512 | 16385 | 1024/512 = 2 |  | 1.882080
256 | 32769 | 1024/256 = 4 |  | 1.533952
128 | 65537 | 1024/128 = 8 |  | 1.346752

## Question 2A

In [17]:
!nvcc HW2_P2.cu -o HW2_P2

In [18]:
!./HW2_P2

**********GPU info**********
Name: Tesla T4
Compute capabili_ty: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835398144 _bytes
****************************

Processing time (use host): 2479.260254 ms

Basic Matrix Multiplication:
Grid size: 32 * 32, block size: 32 * 32
Processing time (use device): 8.028256 ms
Error between device result and host result: 0.000002

Shared memory Matrix Multiplication:
Grid size: 32 * 32, block size: 32 * 32
Processing time (use device): 5.282016 ms
Error between device result and host result: 0.000002

In [ ]:
!nvprof ./HW2_P2

==4039== NVPROF is profiling process 4039, command: ./HW2_P2
**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835398144 bytes
****************************

Processing time (use host): 3827.293945 ms

Basic Matrix Multiplication:
Grid size: 32 * 32, block size: 32 * 32
Processing time (use device): 8.042592 ms
Error between device result and host result: 0.000002

Shared memory Matrix Multiplication:
Grid size: 32 * 32, block size: 32 * 32
Processing time (use device): 5.379104 ms
Error between device result and host result: 0.000002==4039== Profiling application: ./HW2_P2
==4039== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   34.72%  3.2686ms         1  3.2686ms  3.2686ms  3.2686ms  matrix_multiplication_kernel1(float*, float*, float*, int, int, int)
                   28.05%  2.6408ms         1  2.6408ms  2.6408ms  2.6408

**Kernel Optimization:**

The kernel execution times seem reasonable

**Data Transfer Overhead:**

The significant time spent on data transfers suggests that optimizing data movement between the host and device could lead to performance improvements

**Memory Allocation Overhead:**

Memory allocation and deallocation times are relatively low, but depending on the application, optimizing memory management could still provide some benefits.

## Question 2B

**For Basic Matrix Multipication**

1. How many floating operations are being performed in your matrix multiply
kernel? Explain

The number of floating-point operations per element in the resulting matrix (C) is 2 (one multiplication and one addition). Therefore, the total number of floating-point operations in the kernel is
2 × number of elements in C = 2 × m × k

2. How many global memory reads are being performed by your kernel? Explain

In each thread, there are two reads from global memory for matrices A and B. Therefore, the total number of global memory reads in the kernel is 2 × number of threads = 2 × (m × n + n × k)

3. How many global memory writes are being performed by your kernel? Explain.

In each thread, there is one write to global memory for matrix C. Therefore, the total number of global memory writes in the kernel is number of threads = m × k

**For Tiled Matrix Multipication**

1. How many floating operations are being performed in your matrix multiply
kernel? Explain

The number of floating-point operations per element in the resulting matrix (C) is 2 (one multiplication and one addition). Therefore, the total number of floating-point operations in the kernel is m * n * k

2. How many global memory reads are being performed by your kernel? Explain

Each thread reads one element from matrices A and B in shared memory (since shared memory is used for optimization). Therefore, the total number of global memory reads in the kernel is 2 * (n/blocksize) * (m/blocksize) * (k/blocksize)

3. How many global memory writes are being performed by your kernel? Explain

Each thread writes one element to matrix C in global memory. Therefore, the total number of global memory writes in the kernel is number of threads = m × k